### PR TITLE
fix: Use op names for noise parsing

### PR DESCRIPTION
**Pull request by speller26**
Also add license headers

*Issue #, if available:*

*Description of changes:*

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/README.md) and [API docs](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by ajberdy — APPROVED** [review]

**Comment by ajberdy** [comment]
Breaking change because we're adding an abstract method to program context, which is used in the bdk `from_ir` method. We could mitigate this by using `NotImplementedError` instead

**Review by ajberdy — CHANGES_REQUESTED** [review]
We should avoid adding abstract methods that will break the bdk. At least consider temporarily using not implemented exceptions instead

File: src/braket/default_simulator/openqasm/program_context.py [hunk 1]
```diff
@@ -774,16 +798,25 @@ def add_custom_unitary(
...
     @abstractmethod
-    def add_noise_instruction(self, noise: KrausOperation):
+    def add_noise_instruction(
+        self, noise_instruction: str, target: List[int], probabilities: List[float]
+    ):
         """Abstract method to add a noise instruction to the circuit
+
+        Args:
+            noise_instruction (str): The name of the noise operation
+            target (List[str]): The target qubit or qubits to which the noise operation is applied.
```
Comment (Sai-prakash15) [inline src/braket/default_simulator/openqasm/program_context.py:808]:
```suggestion
            target (List[int]): The target qubit or qubits to which the noise operation is applied.
```

File: src/braket/default_simulator/openqasm/program_context.py [hunk 2]
```diff
@@ -774,16 +798,25 @@ def add_custom_unitary(
...
+
         Args:
-            noise (KrausOperation): The noise operation to be added to the circuit
-                - `target` (Tuple[int]): The target qubit or qubits to which the noise operation is applied.
-                - `gamma` (float): The probability of the noise operation occurring. (Optional)
-                - `probability` (float): The probability associated with possible outcome
-                                        of the noise operation. (Optional)
-                - `probabilities` (List[float]): The probabilities associated with each possible outcome
-                                                of the noise operation. (Optional)
+            matrices (List[np.ndarray]): The matrices defining the Kraus operation
```
Comment (Sai-prakash15) [inline src/braket/default_simulator/openqasm/program_context.py:818]:
```suggestion
            matrices (List[ndarray]): The matrices defining the Kraus operation
```

File: src/braket/default_simulator/openqasm/program_context.py [hunk 3]
```diff
@@ -774,16 +798,25 @@ def add_custom_unitary(
...
+            target (List[str]): The target qubit or qubits to which the Kraus operation is applied.
```
Comment (Sai-prakash15) [inline src/braket/default_simulator/openqasm/program_context.py:819]:
```suggestion
            target (List[int]): The target qubit or qubits to which the Kraus operation is applied.
```

**Comment by speller26** [comment]
Veering away from new `abstractmethod`s makes sense; thinking about it further, the only _truly_ required method is `add_gate`, so I'll actually give the other methods the `NotImplemented` treatment.

**Comment by ajberdy** [comment]
Changing the signature for `add_noise_instruction` will be a breaking change for `from_ir`. Maybe we are ok with that since it's brand new code as long as the update in `from_ir` is ready to be merged immediately after

**Review by ajberdy — APPROVED** [review]

**Pull request merged**